bazel_utils.go

```diff
@@ -0,0 +1,182 @@
+	var receiver string
+	sig := obj.Type().(*types.Signature)
+	if recv := sig.Recv(); recv != nil {
+		receiver = recv.Type().String()
+	}
+	// Example receiver: `*github.com/cockroachdb/cockroach/pkg/sql/colfetcher.cFetcher`.
+	// Truncate receiver by removing package path.
+	splitAt := strings.LastIndex(receiver, ".")
+	truncatedReceiver := receiver[splitAt+1:]
+	if strings.HasPrefix(receiver, "*") {
+		// Re-add '*' for pointer receivers such as the example above.
+		return "*" + truncatedReceiver
+	}
+	return truncatedReceiver
+}
+
+func getFnCallIdentifierInObjectFile(obj types.Object) string {
+	fnCallIdentifierInObjectFile := obj.Pkg().Path()
+	receiver := getFnReceiver(obj)
+	if receiver != "" {
+		return fmt.Sprintf("%s.(%s).%s", fnCallIdentifierInObjectFile, receiver, obj.Name())
+	}
+	return fmt.Sprintf("%s.%s", fnCallIdentifierInObjectFile, obj.Name())
+}
+
+// assertInlineByInspectingObjectFiles disassembles machine code and searches
+// through the text symbols for calls to functions that must be inlined.
+func assertInlineByInspectingObjectFiles(w io.Writer, paths []string) error {
+	var bazelBin string
+	var err error
+	if bazelBin, err = getBazelBin(); err != nil {
+		return err
+	}
+	var workspacePath string
+	if workspacePath, err = getWorkspacePath(); err != nil {
+		return err
+	}
+
+	pr, pw := io.Pipe()
+	var f *os.File
+	var mw io.Writer
+	if debugMode {
+		// Create a temp file to log all diagnostic output.
+		var err error
+		f, err = os.CreateTemp("", "gcassert-object-code-inspection*.log")
+		if err != nil {
+			return err
+		}
+		mw = io.MultiWriter(pw, f)
+		defer fmt.Printf("See %s for full output.\n", f.Name())
+	}
+
+	disassembleCmdError := make(chan error, 1)
+	go func() {
+		defer func() {
+			_ = pw.Close()
+			if f != nil {
+				_ = f.Close()
+			}
+		}()
+		for _, path := range paths {
+			getPackageName := func(path string) string {
+				return path[strings.LastIndex(path, "/")+1:]
+			}
+
+			absPath, err := filepath.Abs(path)
+			if err != nil {
+				disassembleCmdError <- err
+				return
+			}
+			relPkgPath, err := filepath.Rel(workspacePath, absPath)
+			if err != nil {
+				disassembleCmdError <- err
+				return
+			}
+			objectFilePath := filepath.Join(bazelBin, relPkgPath, getPackageName(path)+".a")
+			args := []string{"tool", "objdump", objectFilePath}
+			objectFileDisassembleCmd := exec.Command("go", args...)
+			if debugMode {
+				objectFileDisassembleCmd.Stdout = mw
+				objectFileDisassembleCmd.Stderr = mw
+				// Log full 'objdump' command.
+				fmt.Fprintln(f, objectFileDisassembleCmd)
+			} else {
+				objectFileDisassembleCmd.Stdout = pw
+				objectFileDisassembleCmd.Stderr = pw
+			}
+
+			if err := objectFileDisassembleCmd.Run(); err != nil {
+				disassembleCmdError <- err
+				return
+			}
+		}
+		disassembleCmdError <- nil
+	}()
+
+	scanner := bufio.NewScanner(pr)
+
+	// callRegex is used to read a function identifier from a CALL line in the object file.
```

Comment:
> Is this format the same regardless of the CPU architecture?

This works on x86 machines in CI and on my ARM64 M1 machine. 

> I'm a bit concerned that if this format differs between architectures or changes in future Go versions, then gcassert will silently do nothing when run via bazel.

Good point, should I add an assertion that fails if the regex doesn't match any line. Normally it should match too many lines because it's matching all function calls within the paths we are gc-asserting on.

Maybe I should just add a step to this [runbook](https://github.com/cockroachdb/cockroach/tree/master/build#updating-the-golang-version) to make sure `gcassert` is not broken? We always follow this runbook when upgrading Go version.

Reply:
I think I'm satisfied with an assertion / warning you added, and adding extra step to the runbook sgtm but I'll defer to you on that one.